### PR TITLE
Avoid expensive log.Valuer evaluation for disallowed level

### PR DESCRIPTION
**Pull request by damnever**
A lot of failed operations in memcach_client.SetAsync will consume a lot of CPU:
image

```
goos: darwin
goarch: arm64
pkg: github.com/cortexproject/cortex/pkg/util/log
                      │   old.txt    │               new.txt               │
                      │    sec/op    │   sec/op     vs base                │
DisallowedLogLevel-10   1594.5n ± 6%   418.2n ± 1%  -73.77% (p=0.000 n=20)

                      │   old.txt   │              new.txt               │
                      │    B/op     │    B/op     vs base                │
DisallowedLogLevel-10   1567.0 ± 0%   623.0 ± 0%  -60.24% (p=0.000 n=20)

                      │  old.txt   │              new.txt               │
                      │ allocs/op  │ allocs/op   vs base                │
DisallowedLogLevel-10   23.00 ± 0%   13.00 ± 0%  -43.48% (p=0.000 n=20)
```

**Checklist**
- [x] Tests updated
- [ ] Documentation added
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

File: pkg/util/log/log.go [hunk 1]
```diff
@@ -72,24 +69,32 @@ type PrometheusLogger struct {
 // NewPrometheusLogger creates a new instance of PrometheusLogger which exposes
 // Prometheus counters for various log levels.
 func NewPrometheusLogger(l logging.Level, format logging.Format) (log.Logger, error) {
```
Comment (alanprot) [inline pkg/util/log/log.go:71]: This is not being used anymore?

**Comment by alanprot** [comment]
Can you explain what is the functional change here? are we missing/changing something?

File: pkg/util/log/log.go (hunk 1 above)
Comment (damnever) [reply pkg/util/log/log.go:71]: This is used by other packages.

**Review by alanprot — APPROVED** [review]

**Review by alanprot — APPROVED** [review]

**Review by alvinlin123 — APPROVED** [review]
This is an awesome enhancement, thank you so much!

**Pull request merged**